File: Solution/BOJ/2042 구간 합 구하기.cpp
#include <iostream>

typedef long long ll;
using namespace std;
int N, M, K, S;

long makeTree(int node, int left, int right, ll arr[], ll tree[]) {
	if (left == right) {
		if (left <= N) {
			return tree[node] = arr[left];	
		} else {
			return tree[node] = 0;
		}
	}
	
	int mid = (left + right) / 2;
	tree[node] = makeTree(node * 2, left, mid, arr, tree);
	tree[node] += makeTree(node * 2 + 1, mid + 1, right, arr, tree);
	
	return tree[node];
}

long query(int node, int left, int right, int qLeft, int qRight, ll tree[]) {
	if (right < qLeft || qRight < left) {
		return 0;
	} else if (qLeft <= left && right <= qRight) {
		return tree[node];
	} else {
		int mid = (left + right) / 2;
		return query(node * 2, left, mid, qLeft, qRight, tree) + query(node * 2 + 1, mid + 1, right, qLeft, qRight, tree);
	}
}

void update(int node, int left, int right, int index, ll diff, ll tree[]) {
	if (left <= index && index <= right) {
		tree[node] += diff;
		if (left != right) {
			int mid = (left + right) / 2;
			update(node * 2, left, mid, index, diff, tree);
			update(node * 2 + 1, mid + 1, right, index, diff, tree);
		}
	}
}

int main() {
	cin >> N >> M >> K;
	ll arr[N+1];
	for (int i = 1; i <= N; i++) {
		cin >> arr[i];
	}
	S = 1;
	while (S < N) {
		S *= 2;
	}
	ll tree[S * 2];
	makeTree(1, 1, S, arr, tree);
	for (int i = 0; i < M + K; i++) {
		int a, b, c;
		cin >> a >> b >> c;
		if (a == 1) {
			update(1, 1, S, b, c - arr[b], tree);
		} else {
			cout << query(1, 1, S, b, c, tree) << endl;
		}
	}
}

File: Solution/BOJ/2042 구간 합 구하기.py
import sys

input = sys.stdin.readline


def update(i, diff):
    while i <= n:
        tree[i] += diff
        i += i & -i


def prefix_sum(i):
    result = 0
    while i > 0:
        result += tree[i]
        i -= i & -i

    return result


def interval_sum(s, e):
    return prefix_sum(e) - prefix_sum(s - 1)


n, m, k = map(int, input().split())
arr = [0] * (n + 1)
tree = [0] * (n + 1)
for i in range(1, n + 1):
    x = int(input())
    arr[i] = x
    update(i, x)


for _ in range(m + k):
    a, b, c = map(int, input().split())
    if a == 1:
        update(b, c - arr[b])
        arr[b] = c
    else:
        print(interval_sum(b, c))
